src/behavior-considered-undefined.md

```diff
@@ -151,10 +155,30 @@ entire range, so it is important that the length metadata is never too large. In
 particular, the dynamic size of a Rust value (as determined by `size_of_val`)
 must never exceed `isize::MAX`.
 
+### Unwinding inappropriately
+[unwinding]: #unwinding-inappropriately
+
+Unwinding with the wrong [ABI][abi] is undefined behavior:
+
+* Causing an unwind into Rust code from a foreign function that was called via a
+  function declaration or pointer declared with a non-unwinding ABI, such as `"C"`,
+  `"system"`, etc. (For example, this case occurs when such a function written in
+  C++ throws an exception that is uncaught and propagates to Rust.)
+* Calling a Rust `extern` function that unwinds (with `extern "C-unwind"` or
+  another ABI that permits unwinding) from a runtime that does not support.
+  unwinding, such as code compiled with GCC or Clang using `-fno-exceptions`
+
+Catching an unwinding operation in a different runtime than the one in which it
+initiated is also undefined behavior:
+
+* Catching a Rust `panic` in non Rust code (for instance `catch (...)` in C++).
```

Comment:
This isn't even UB in Rust, it's UB in a different language. I don't think that belongs here.

It may be a good idea to describe the UB concerns that come up with FFI, but that would be a new page in the reference. Those concerns go well beyond unwinding.

Reply:
My understanding is that it can cause Rust UB, though. For instance, the global atomic panic count might not be updated correctly. @Amanieu, @chorman0773, @nikomatsakis, what do you think about removing this bullet?

Reply:
I'm mixed here - on one hand, it's not Itanium level UB in C++ (though rethrowing it is), so if it causes UB, it has to be Rust UB. Arguably it's strictly UB in the C++ Standard, as the standard is wholly silent on the idea of other languages that aren't C, but C++ implementations implement specific supersets of C++, and I believe FFI-unwind guarantees compatibilty with them. 

On the other hand, it does seem like "Not Our Problem:tm:".

Reply:
My personal stance is that "it's not our problem" is an okay reason not to put effort into writing explicit guarantees, but it seems like a very weird reason to remove what I've already written here. I don't see how it could cause extra confusion or other problems.

Reply:
Putting T-spec-contributor hat on, I'd agree - I don't think there is a pressing need to remove the text (though other T-spec folks might have different opinions).

Reply:
I find it strange because it is qualitatively different UB than everything else on this page. Everything else we'll be able to point to a line of MiniRust checking the corresponding condition. But for this one, that doesn't even make any sense.

I think we shouldn't mix up "UB in the Rust Abstract machine" with "UB as part of rustc's implementation of the Rust AM on actual hardware". This page, so far, is only about the former. Now you're adding the latter somewhere in the middle of it and I think that's a bad mixing of concerns.

I'm not necessarily opposed to having them both on this page, but they should be very clearly separate IMO.

Reply:
Perhaps we can phrase it with something requirement like "An unwind originated from a Rust runtime must either lead to termination or be caught by the same runtime."

Reply:
@RalfJung I'm sympathetic to the idea of wanting there to be a clean line dividing those types of UB, but I'm not quite convinced that that's an existing characteristic of the Reference that I'm losing here. In any case, breaking UB in the Reference into these two explicit categories is something I'd be happy to try to help with, but not as part of this PR, if I can avoid it (because of time pressure - I think landing a change to the reference as close as possible to the stabilization of the change in `extern "C"` behavior is highly desirable, and in all fairness I should have done this back when we stabilized `extern "C-unwind"`).

In any case, I've rephrased it to be much more clearly about Rust behavior, in my opinion:

> There are currently no guarantees about the behavior that occurs when a foreign runtime attempts to dispose of, or rethrow, a Rust `panic` payload. In other words, an unwind originated from a Rust runtime must either lead to termination of the process or be caught by the same runtime.

Reply:
> I'm not quite convinced that that's an existing characteristic of the Reference that I'm losing here. 

Currently, as far as I can tell everything on this page has a direct correspondence with an Abstract Machine error condition. That is an existing characteristic of this page in the reference. If you disagree with that claim, please name a counterexample. You are changing the page in a way that it loses that characteristic. The very mention of "foreign runtime" loses that characteristic.

I can only repeat: If we start spelling out the assumptions that the Rust *runtime* makes (which are an implementation detail of how rustc implements the Abstract Machine), that's going to be a very long list. Unwinding is but one of many aspects, and only discussing that one aspect here is entirely arbitrary.

I see that we have a page `runtime.md` but it just discusses some attributes? That's quite different from what we are talking about.

I would also like to see C-unwind stabilizing ASAP, but I am still not happy with this change to `behavior-considered-undefined.md`. So here's a counter-proposal:
- We add a new item to the UB list:
```
- Violating assumptions of the Rust runtime. This is only possible using mechanisms outside
  Rust. Most assumptions of the Rust runtime are currently not explicitly documented. For
  assumptions specifically related to unwinding, see [...].
```
- This item then links to some page that explains whatever you this it would be good for the reference to explain about unwinding -- probably a subsection of `panic.md`?

Reply:
Please see what you think now.

Reply:
Yeah I  like this much better, thanks! Will leave comments for concrete questions.